### ISSUE TITLE
[HOLD for payment 2024-02-26] [$500] Country search does not consider Latin or Spanish letters in search

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#29826]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.85-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1697544511147309

## Action Performed:
1. Open the app
2. Open settings->profile->personal details->address->country
3. Search with letter 'Å' and observe that even though we have country with that letter in list, app still doesn't show that on n top of search eg: 'Åland Islands'
4. Similarily search with letter 'é', we have multiple countries with that letter but search doesn't show them on top eg: 'São Tomé & Príncipe', 'Saint Barthélemy'
## Expected Result: 
App should consider latin or spanish letter in country search and display their results on top as we generally do on user search page and send message page
## Actual Result: 
App does not consider latin or spanish letter in country search and does not display their results on top
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/38435837/ea1fd0e6-d2ae-4ddd-ae9f-009e029ddbbc

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/38435837/8bb9a14b-5355-4f92-b97a-288adbeaed1b

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/38435837/adb46382-7332-407f-8d6d-54a484660f9d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/38435837/1537c7ac-d8cc-4472-862a-3e2c63c78dde

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/38435837/0ded0b82-72cd-4cca-93c4-59eb0eff0779

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/38435837/f35a61db-8c14-4277-94b9-d3d539d308a9

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c62c8f44bacecbe2</li>
        <li>Upwork Job ID: 1714381091711156224</li>
        <li>Last Price Increase: 2023-10-31</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Reviewer | 27762279</li>
<li>graylewis | Contributor | 27762281</li>
<li>dhanashree-sawant | Reporter | 27762282</li>
</ul></li>
    </ul>
</details>

**Comment by graylewis** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
The country search page doesn't properly search using non-latin characters

### What is the root of the issue
Items in the array of countries being searched get sanitized with the StringUtils.sanitizeString() function.

### What changes do you think we should make in order to solve the problem?
Create a function which sanitizes strings without stripping them of their non-latin characters 

What alternative solutions did you explore? (Optional)
NA

**Comment by Victor-Nyagudi** [comment]
It seems this method removes letters with diacritics i.e. accented letters like é, so that could be affecting country search results. 

https://github.com/Expensify/App/blob/1c051fe49a9e4423053c1953c36dacd3ee81c8c0/src/libs/StringUtils.ts#L4-L11

I'm not sure if this is the intended behavior or a regression, so [I asked about it](https://github.com/Expensify/App/pull/24477/files#r1363945851) in the pull request that introduced this method. 

- https://github.com/Expensify/App/pull/24477

**Comment by strepanier03** [comment]
@situchan - Friendly bump on the proposal above.

**Comment by situchan** [comment]
@graylewis your solution will cause regression - https://github.com/Expensify/App/issues/24344
Please try search `cote d'Ivoire`

**Comment by graylewis** [comment]
@situchan Thanks for the heads up! Editing my original proposal now with a more robust solution

**Comment by strepanier03** [comment]
Friendly bump @situchan for the updated proposal from @graylewis. Thanks!

**Comment by situchan** [comment]
Looking for better solution

**Comment by graylewis** [comment]
@situchan Can I ask what you feel is lacking in my proposed solution?

**Comment by situchan** [comment]
Waiting for confirmation on https://github.com/Expensify/App/pull/24477/files#r1363945851

**Comment by strepanier03** [comment]
All good here Melvin.

**Comment by situchan** [comment]
No response yet from linked PR

🎀 👀 🎀 to discuss the expected behavior with engineer

**Comment by situchan** [comment]
@francoisl thanks for being assigned here. Do you agree the results exactly matching keyword should be on top?

<img width="372" alt="Screenshot 2023-11-02 at 4 40 01 PM" src="https://github.com/Expensify/App/assets/108292595/6443ee8e-a452-4318-8185-ea84237a5367">


This is now alphabetical order.

This issue is trying to put `Åland Islands` above `Afghanistan` because `Afghanistan` doesn't contain `Å`

**Comment by francoisl** [comment]
Yes I agree - if you make the extra effort to enter diacritics in your search value, it makes sense that we'd take them into account. I'd say the ideal sorting algorithm would:

- Take diacritics into account (so if I search for `Å`, then `Åland Islands` would be the first result)
    - Note: if you _don't_ use diacritics, still include country names that contain them, like we do now. So if you search for `cote`, we'd still include `Côte d'Ivoire`

- Prioritize countries whose name _starts_ with your search value, rather than merely _contains_ it
    - For example, if you search for `bar`, at the moment the sorted country list is:
        1. `Antigua & Barbuda`
        2. `Barbados`
        <img width="383" alt="image" src="https://github.com/Expensify/App/assets/2229301/93c058bb-f94f-485d-bb5f-0cac09eac1be">
    I feel like it would make more sense to have **Bar**bados first in this case?

I'm open to other suggestions or curious if anyone disagrees though!

**Comment by strepanier03** [comment]
all good here Melvin, discussion is happening and moving things forward.

**Comment by strepanier03** [comment]
All good Melvin, things are moving forward.

**Comment by francoisl** [comment]
@graylewis does [this](https://github.com/Expensify/App/issues/29826#issuecomment-1791346896) make sense? Let us know if you feel like updating your proposal!

**Comment by graylewis** [comment]
@francoisl yep, I agree completely. I'll update my proposal in the next few hours to reflect the new requirements

**Comment by francoisl** [comment]
Thanks @graylewis. @situchan can you review the updated proposal when you have a chance please?

**Comment by situchan** [comment]
reviewing proposal

**Comment by situchan** [comment]
@graylewis I am not convinced about your pseudo code.

`a.searchValue.toLowerCase().substring(2, trimmedSearchValue.length + 2)` - can you explain detail?

Please try to explain in plain English:
- First, sort by ...
- Second, sort by ...
- ...

**Comment by graylewis** [comment]
@situchan thanks for taking the time to review my proposal. It's actually not pseudo-code, I've run the code in the app to make sure it works. I'll paste the summary from my proposal and also try to summarize it for clarity.

1. Data is filtered based on the original search term comparison. This includes all possible matches.
2. A sort is run to prioritize exact matches from the beginning of the country name, as requested in the updated requirements. E.g. "Bar" returns Barbados first since it's the beginning of the country name.
3. The sanitized search term is compared to the raw search term. If they're different, then diacritics/non-latin chars are present in the search term. In this case, exact matches including diacritics are pushed to the top.
4. If diacritics aren't present, then as is done in the existing code, we prioritize exact country code matches. E.g. "US" returns the united states before anything else. (Since country codes never contain diacritics, there's no need to run this step if diacritics are in the search term)

In summary: 
1. First the data is filtered for all possible matches.
2. Then, a sort is run which prioritizes exact matches from the start of the word. (this is one of the revised requirements from @francoisl)
3. Then, the sanitized search term is compared to the unsanitized search term. If they're different, then diacritics were present in the search term. If this is the case, then the exact matches containing the diacritic are sorted to the top.
4. Otherwise, exact country code matches e.g. US for the united states, are prioritized to the top.

**Comment by situchan** [comment]
@graylewis's [general idea](https://github.com/Expensify/App/issues/29826#issuecomment-1767156275) looks good to me.
Expected behavior for clarification: https://github.com/Expensify/App/issues/29826#issuecomment-1791346896
Also, make sure to add unit test with various test cases.
🎀 👀 🎀 C+ reviewed

**Comment by graylewis** [comment]
@MelvinBot almost done with the PR, just writing some unit tests

**Comment by strepanier03** [comment]
Thank you @graylewis.

**Comment by francoisl** [comment]
Hi @graylewis, how are things going with the PR?

**Comment by graylewis** [comment]
Hi @francoisl, I'm done with the actual code and I have an almost complete draft PR, I've just been trying to get the screen recordings for all the different platforms done and I had to do some reconfiguring to get the iOS virtual device running properly etc. I'm expecting to submit the PR tomorrow

**Comment by francoisl** [comment]
Sounds good, thanks for the update!

**Comment by situchan** [comment]
@graylewis can we get update on [PR](https://github.com/Expensify/App/pull/32151)?

**Comment by graylewis** [comment]
@situchan sorry for the wait, didn't see that you'd reviewed the PR. made the requested changes to the PR

**Comment by situchan** [comment]
> @situchan sorry for the wait, didn't see that you'd reviewed the PR. made the requested changes to the PR

@graylewis https://github.com/Expensify/App/pull/32151#issuecomment-1913611622

**Comment by strepanier03** [comment]
@graylewis and @dhanashree-sawant - I've paid you both via Upwork and closed the contracts. 

@situchan - I'll check in later today to see if the checklist is done and if so move forward then. If not, I'll check tomorrow morning.

**Comment by situchan** [comment]
This is not really bug but improvement. We added enough automated tests so no need regression test.

**Comment by strepanier03** [comment]
Got it thanks for the context @situchan - I'll handle now.

**Comment by strepanier03** [comment]
Okay, payment sent and contract closed. 

Thanks again everyone 👏
